### PR TITLE
(1.18.x) Add multiloader registration system

### PR DESCRIPTION
**Pull request by Matyrobbrt**
This PR proposes a new registry system, which does **not** rely on classloading, and that is guaranteed to always work.
The registration system uses ServiceLoaders in order to let each platform register objects differently.

The usage is somewhat similar to Forge's `DeferredRegister`, as that is what is being used internally on the Forge side, and direct immediate `Registry#register` calls on Fabric. 

Example usage:
```java
public static final RegistrationProvider<Test> PROVIDER = RegistrationProvider.get(Test.REGISTRY, "modid");
public static final RegistryObject<Test> OBJECT = PROVIDER.register("object", () -> new Test());

/** The purpose of this method is to be called in the mod's constructor, in order to assure that the class is loaded, and that objects can be registered. */
public static void loadClass() {}
```

### CODE REVIEW

File: Fabric/src/main/java/potionstudios/byg/FabricRegistrationFactory.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+                }
+
+                @Override
+                public I get() {
+                    return obj;
+                }
+
+                @Override
+                public Holder<I> asHolder() {
+                    return (Holder<I>) registry.getOrCreateHolder(ResourceKey.create(ResourceKeyAccess.create(registry.key().location(), rl), rl));
```
Comment (CorgiTaco) [inline Fabric/src/main/java/potionstudios/byg/FabricRegistrationFactory.java]: should return `(Holder<I>) registry.getOrCreateHolder((ResourceKey<T>) this.key)` in order for ConfiguredFeature Holder's to function on fabric.

File: Fabric/src/main/java/potionstudios/byg/client/BYGFabricClient.java [hunk 2]
```diff
@@ -16,7 +18,13 @@ public void onInitializeClient() {
...
 
-        BYG.clientLoad();
+        BYGClient.load();
+        BYGClient.threadSafeLoad();
         BYGCutoutRenders.renderCutOuts(blockRenderTypeMap -> ItemBlockRenderTypeAccess.byg_getTYPE_BY_BLOCK().putAll(blockRenderTypeMap));
+
+        BYGEntityRenderers.register(AccessEntityRenderers::byg_register);
+
+        final var particleAccessor = ((ParticleEngineAccess) Minecraft.getInstance().particleEngine);
+        BYGClient.registerParticles(particleAccessor::byg_register);
```
Comment (CorgiTaco) [inline Fabric/src/main/java/potionstudios/byg/client/BYGFabricClient.java]: Throws NPE because Particle Engine is null

**Review by CorgiTaco — CHANGES_REQUESTED** [review]
Fix Fabric Game Load, ensure fabric/forge servers run also.

**Comment by Matyrobbrt** [comment]
All errors have been solved, Fabric & Forge client & servers have been tested.

**Review by CorgiTaco — APPROVED** [review]

**Pull request merged**